pkg/apis/networking/types.go

```diff
@@ -235,12 +235,25 @@ type IngressList struct {
 
 // IngressSpec describes the Ingress the user wishes to exist.
 type IngressSpec struct {
-	// A default backend capable of servicing requests that don't match any
-	// rule. At least one of 'backend' or 'rules' must be specified. This field
+	// Class is the name of the IngressClass cluster resource. The associated
+	// IngressClass defines which controller will implement the resource.
+	// This replaces the deprecated `kubernetes.io/ingress.class` annotation.
+	// For backwards compatibility, when that annotation is set, it must be
+	// given precedence over this field. The controller may emit a warning if
+	// the field and annotation have different values. Implementations of this
+	// API may choose to ignore Ingresses without a class specified.
+	// An IngressClass resource may be marked as default, which can be used to
```

Comment:
we should be more specific concerning marking as default. We don't want to be mysterious here, people will have a hard time finding it.

Reply:
@bowei I have more specific documentation on the IngressClass resource - it was missing in one of the versions when you reviewed this. I was hoping to avoid duplicating that copy on both the Class field and the Class resource, but I can do that if you think it will help make this easier to understand.

@hbagdi This was a last minute addition to the KEP per Tim's suggestion. It mirrors the behavior of StorageClass. An admission controller watches Ingresses and sets a default value for Class if it's not already specified on the Ingress and there is an IngressClass resource that has been marked as default.

Reply:
I'm not saying the addition doesn't make sense and the behavior is incorrect.

But, this:

> Implementations of this API may choose to ignore Ingresses without a class specified.

This seems like a problematic `may` to me.
We should define the behavior of controllers around what to do when the class is not set at all. If one controller ignores them, and user installs another one, and it takes over those Ingress resources, then that leads to a bad user experience. The guidelines here should be more strict.

Reply:
Yeah, I agree that it would be nice if we could be more specific here. Unfortunately we're limited to some extent by what already exists. Ingresses have already been valid without a class specified, and we can't make that use case invalid and still maintain backwards compatibility. At the same time, the behavior when this was unspecified was also not specified and therefore was inconsistent, so we're limited with what we can say here without requiring behavior to change dramatically here. 

If there were a way to require class to be specified here, that sure would be ideal, but I just don't think we can unfortunately. I'm also open to ideas for how we could further specify how the lack of a class should be handled, I just haven't found anything that works well within the constraints we have (especially backwards compatibility).

Reply:
However, this is a V1 spec for which (AFAIK) every ingress controller will need code changes to support (even if it is just adding informers for the new types). This seems like the best opportunity to make another incompatible change :)

Reply:
James, the problem is that Ingress has been v1beta1 for so long that is essentially a GA API, and beta is only in the name. People rely on it like a v1 GA API and breaking the behavior will not go well for most users of the community.

Rob, you have a good point.
I get that we can't enforce things in the API but we can have stricter guidelines.

A solution I can think of is to replace `may` with `should` and then detail on the behavior we want implementors to follow.

Reply:
:+1: I like that idea, thanks! I've replaced the `may` with `should` and I think that at least gets us closer to the goal here.